### PR TITLE
Feat: add passthrough to genesis

### PR DESCRIPTION
**Pull request by mustermeiszer**
# Description
Currently, we always need to have real Axelar routers on testnets. That prevents us from texting LP logic on development for example. 

This PR adds a passthrough router on the EVM side to all new chains. The router always sits at the same location (`0x33e7daf228e7613ba85ef6c3647dbceb0f011f7c`) and is based on [this](https://github.com/centrifuge/liquidity-pools/blob/d3da102e7bf656fd50feeb888e17f423317aeeb3/test/integration/PassthroughRouter.sol) contract. 

The router simply forwards LP messages to the precompile, without checking anything. The `AxelarPrecompile` of course needs to have `0x33e7daf228e7613ba85ef6c3647dbceb0f011f7c` set as its trusted `GatewayContract`. 

## Feat
*  Enables testing LP logic on development - making development our main testground and demo will move more to a public testground for external users

## Changes and Descriptions
* Adds `runtime_common::evm::utils::account_codes()` - generates the genesis codes for the EVM side 
* Adds passthrough router bytecode as constant
* Adds passthrough router `H160` by deriving a hash from a constant input

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: runtime/common/src/evm/mod.rs [hunk 1]
```diff
@@ -107,3 +107,55 @@ impl<T: pallet_aura::Config<AuthorityId = AuraId>> FindAuthor<H160> for FindAuth
 		None
 	}
 }
+
+/// Passthrough router deployed bytecode as of this state https://github.com/centrifuge/liquidity-pools/tree/d3da102e7bf656fd50feeb888e17f423317aeeb3
```
Comment (wischli) [inline runtime/common/src/evm/mod.rs:111]: Q: Which code changes on solidity side lead to a required update of the bytecode?

**Review by wischli — DISMISSED** [review]
Great improvement!

File: runtime/common/src/evm/mod.rs (hunk 1 above)
Comment (mustermeiszer) [reply runtime/common/src/evm/mod.rs:111]: Just the ones in the file linked above. Should maybe make that clear in the comment.

File: node/src/chain_spec.rs [hunk 2]
```diff
@@ -352,7 +352,7 @@ fn centrifuge_genesis(
 			"chainId": Into::<u32>::into(chain_id),
 		},
 		"evm": {
-			"accounts": runtime_common::evm::precompile::utils::precompile_account_genesis::<centrifuge_runtime::Precompiles>(),
+			"accounts": runtime_common::evm::utils::account_genesis::<centrifuge_runtime::Precompiles>(),
```
Comment (lemunozm) [inline node/src/chain_spec.rs:355]: Do we also change this for testing purposes right?

**Review by lemunozm — APPROVED** [review]
LGTM! This change does not imply any change/simplification in the current `lp` tests?

File: node/src/chain_spec.rs (hunk 2 above)
Comment (mustermeiszer) [reply node/src/chain_spec.rs:355]: Yes.

Comment (mustermeiszer) [reply node/src/chain_spec.rs:355]: @wischli we should probably also generate a new dev genesis, right? For the life environment, or don't we use a static one?

File: runtime/common/src/evm/mod.rs [hunk 3]
```diff
@@ -107,3 +110,76 @@ impl<T: pallet_aura::Config<AuthorityId = AuraId>> FindAuthor<H160> for FindAuth
...
+
+/// Input for the KeccakHasher to derive a random `H160` where the passthrough
+/// router is always located at. Refers to address:
+/// `0x33e7daf228e7613ba85ef6c3647dbceb0f011f7c`
+const PASSTHROUGH_ROUTER_ACCOUNT_CODES_ACCOUNT_LOCATION_SALT: &[u8] =
+	b"PASSTHROUGH_ROUTER_ACCOUNT_CODES_ACCOUNT_LOCATION_SALT";
+
+pub fn passthrough_router_location() -> H160 {
+	H160::from(sp_core::H256::from(sp_core::KeccakHasher::hash(
```
Comment (lemunozm) [inline runtime/common/src/evm/mod.rs:130]: Curious, who demands the usage of `KeccakHasher`?

**Review by lemunozm — DISMISSED** [review]
LGTM

**Comment by wischli** [comment]
@mustermeiszer Sorry for being intrusive. I can imagine you have tons of stuff on your plate before holiday so I made some minor changes to make CI green.

**Review by lemunozm — DISMISSED** [review]

File: runtime/common/src/evm/mod.rs [hunk 4]
```diff
@@ -107,3 +107,76 @@ impl<T: pallet_aura::Config<AuthorityId = AuraId>> FindAuthor<H160> for FindAuth
...
+			passthrough_router_location().as_bytes(),
+			hex_literal::hex!("33e7daf228e7613ba85ef6c3647dbceb0f011f7c")
+		);
+	}
+
+	#[test]
+	fn stable_passthrough_bytecode_hash() {
+		assert_eq!(
+			BlakeTwo256::hash_of(&PASSTHROUGH_ROUTER_ACCOUNT_CODES.to_vec()),
+			hex_literal::hex!("545a48d6f7f1a01cb2bd090a5272cd52c54eafea762071ec652c8ac94610146e")
```
Comment (wischli) [inline runtime/common/src/evm/mod.rs:162]: @mustermeiszer This test fails. I suppose due to changes to the `PassThroughAdapter.sol` between your referenced commit `ee35da785f2af3303ef9f52c6ecefeb9671aa61b` and current main at `223a0f36edabc675f8c74c47b20e366178df7ca3`, see [diff](https://github.com/centrifuge/liquidity-pools/compare/ee35da785f2af3303ef9f52c6ecefeb9671aa61b). 

I wonder how you derived the passthrough account byte codes.

Error
```
thread 'evm::tests::stable_passthrough_bytecode_hash' panicked at runtime/common/src/evm/mod.rs:160:9:
assertion `left == right` failed
  left: 0x31173f15567854cfc3702aa6b639bf0dedf74638e745a3e90fa00f1619d8b94c
 right: 0x545a48d6f7f1a01cb2bd090a5272cd52c54eafea762071ec652c8ac94610146e
```

Comment (mustermeiszer) [reply runtime/common/src/evm/mod.rs:162]: As I can not compile it wanted to retrieve the new hash through the UI. Haha. So using the hash here generated will be okay

Comment (wischli) [reply runtime/common/src/evm/mod.rs:162]: But how can I derive the `PASSTHROUGH_ROUTER_ACCOUNT_CODES`?

Comment (wischli) [reply runtime/common/src/evm/mod.rs:162]: But then I only adjust the blake2 256 hash of the `PASSTHROUGH_ROUTER_ACCOUNT_CODES` but not any of the 3289 bytes of `PASSTHROUGH_ROUTER_ACCOUNT_CODES`. Based on your comments, those need to be adjusted as well?

```
/// Passthrough router deployed bytecode as of this state https://github.com/centrifuge/liquidity-pools/blob/ee35da785f2af3303ef9f52c6ecefeb9671aa61b/test/integration/PassthroughAdapter.sol
///
/// NOTE: If the above file changes, this code needs to be adapted.
...
```

Comment (wischli) [reply runtime/common/src/evm/mod.rs]: Does this change suffice @mustermeiszer ? [949bdff](https://github.com/centrifuge/centrifuge-chain/pull/1888/commits/949bdff2fbd5340560262d6f9343e780cc354573)

Comment (mustermeiszer) [reply runtime/common/src/evm/mod.rs]: Yes, that was, what I intended to do! Perfect! We always need that hash for setting up the routers, so I wanted to have it somewhere ^^

**Review by wischli — APPROVED** [review]

**Pull request merged**